internal/multistorage/cache/status_cache.go

```diff
@@ -21,169 +23,281 @@ type statusCache struct {
+			storageNames(relevant),
+		)
+		checkResult := c.checker.checkForAlive(relevant...)
+		newFile = updateFileContent(newFile, checkResult)
+		allAlive = newFile.getAllAlive(c.storagesInOrder)
 	}
 
-	memCache = newFile
-	allAlive := memCache.getAllAlive(c.storagesInOrder)
-	tracelog.InfoLogger.Printf("Found alive storages: %v", storageNames(allAlive))
 	return allAlive, nil
 }
 
-func (c *statusCache) FirstAliveStorage() (*NamedFolder, error) {
-	memCacheMu.Lock()
-	defer memCacheMu.Unlock()
+func (c *statusCache) FirstAliveStorage() (firstAlive *NamedFolder, err error) {
+	c.sharedMemMu.Lock()
+	defer c.sharedMemMu.Unlock()
 
-	memFirstAlive, allRelevant := memCache.getRelevantFirstAlive(c.ttl, c.storagesInOrder)
+	memFirstAlive, allRelevant := c.sharedMem.getRelevantFirstAlive(c.ttl, c.storagesInOrder)
 	if memFirstAlive != nil {
 		return memFirstAlive, nil
 	}
 	if allRelevant {
-		tracelog.InfoLogger.Print("There is no alive storages in memory cache")
-		return nil, nil
+		tracelog.InfoLogger.Print("There are no alive storages in memory cache")
 	}
 
-	oldFile, err := readFile()
-	if err != nil {
-		tracelog.WarningLogger.Printf("Failed to read cache file, it will be overwritten: %v", err)
+	var oldFile storageStatuses
+	if c.sharedFileUse {
+		oldFile, err = readFile(c.sharedFile)
+		if err != nil {
+			tracelog.WarningLogger.Printf("Failed to read cache file, it will be overwritten: %v", err)
+		}
 	}
 	fileFirstAlive, allRelevant := oldFile.getRelevantFirstAlive(c.ttl, c.storagesInOrder)
 	if fileFirstAlive != nil {
 		tracelog.InfoLogger.Printf("Take first alive storage from file cache: %s", fileFirstAlive.Name)
-		memCache[fileFirstAlive.Key] = oldFile[fileFirstAlive.Key]
+		(*c.sharedMem)[fileFirstAlive.Key] = oldFile[fileFirstAlive.Key]
 		return fileFirstAlive, nil
 	}
 	if allRelevant {
-		tracelog.InfoLogger.Print("There is no alive storages in file cache")
-		return nil, nil
+		tracelog.InfoLogger.Print("There are no alive storages in file cache")
 	}
 
-	_, outdated := oldFile.splitByRelevance(c.ttl, c.storagesInOrder)
+	relevant, outdated := oldFile.splitByRelevance(c.ttl, c.storagesInOrder)
 
-	tracelog.InfoLogger.Printf("Storage status cache is outdated, checking for alive again: %v", storageNames(outdated))
+	newFile := oldFile
+	if len(outdated) > 0 {
+		tracelog.InfoLogger.Printf("Storage status cache is outdated, checking for alive again: %v", storageNames(outdated))
+		checkResult := c.checker.checkForAlive(outdated...)
+		newFile = updateFileContent(newFile, checkResult)
+		firstAlive, _ = newFile.getRelevantFirstAlive(time.Duration(math.MaxInt64), c.storagesInOrder)
+	}
 
-	checkResult := c.checker.checkForAlive(outdated...)
+	defer func() {
+		if c.sharedFileUse {
+			err := writeFile(c.sharedFile, newFile)
+			if err != nil {
+				tracelog.WarningLogger.Printf(
+					"Failed to write cache file, each subsequent command will check the storages again: %v",
+					err,
+				)
+			}
+		}
```

Comment:
let's wrap this logic into some separate func, it will be easier to read and remove the code duplication

Reply:
Done